`berkeley`: Dagster job `ensure_alldocs` fails with `AssertionError`

Today, I visited the Dagit instance in the Berkeley environment (`nmdc-berkeley` namespace on Spin) and tried running the `ensure_alldocs` job.

While the `materialize_alldocs` op was running, an error occurred. Here's a screenshot of the error message, followed by a copy/paste of the same error message:

![image](https://github.com/user-attachments/assets/70b611d8-4bf6-46ce-ab1e-34c7f5c03c93)

<details>
  <summary>
    Show/hide copy/pasted error message
  </summary>

```py
dagster._core.errors.DagsterExecutionStepExecutionError: Error occurred while executing op "materialize_alldocs":

  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_plan.py", line 247, in dagster_event_sequence_for_step
    for step_event in check.generator(step_events):
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_step.py", line 500, in core_dagster_event_sequence_for_step
    for user_event in _step_output_error_checked_user_event_sequence(
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_step.py", line 184, in _step_output_error_checked_user_event_sequence
    for user_event in user_event_sequence:
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/execute_step.py", line 88, in _process_asset_results_to_events
    for user_event in user_event_sequence:
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/compute.py", line 198, in execute_core_compute
    for step_output in _yield_compute_results(step_context, inputs, compute_fn, compute_context):
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/compute.py", line 167, in _yield_compute_results
    for event in iterate_with_context(
  File "/usr/local/lib/python3.10/site-packages/dagster/_utils/__init__.py", line 476, in iterate_with_context
    with context_fn():
  File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/utils.py", line 84, in op_execution_error_boundary
    raise error_cls(

The above exception was caused by the following exception:
AssertionError: configuration_set collection has class name of ['ChromatographyConfiguration', 'MassSpectrometryConfiguration'] and len 2

  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/utils.py", line 54, in op_execution_error_boundary
    yield
  File "/usr/local/lib/python3.10/site-packages/dagster/_utils/__init__.py", line 478, in iterate_with_context
    next_output = next(iterator)
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/compute_generator.py", line 141, in _coerce_op_compute_fn_to_iterator
    result = invoke_compute_fn(
  File "/usr/local/lib/python3.10/site-packages/dagster/_core/execution/plan/compute_generator.py", line 129, in invoke_compute_fn
    return fn(context, **args_to_pass) if context_arg_provided else fn(**args_to_pass)
  File "/opt/dagster/lib/nmdc_runtime/site/ops.py", line 1027, in materialize_alldocs
    len(collection_name_to_class_names[name]) == 1
```
</details>

I think this was the first time that job had been run in the Berkeley environment. That is based upon what I see here, on the "Runs" page of Dagit:

![image](https://github.com/user-attachments/assets/47c12ad2-43ae-468f-b9ca-cbbd7ea376ca)

### Task

The task here is to make it so the `ensure_alldocs` job runs and an alldocs collection exists in the Berkeley Mongo database.

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1080,7 +1080,7 @@ def materialize_alldocs(context) -> int:
             #       determining the class ancestry of only _one_ of them (we call this the "representative") and then
             #       (later) attributing that class ancestry to all of them.
             #
-            representative_doc = next(docs_having_type, default=None)
+            representative_doc = next(docs_having_type)
 
             # Instantiate the Python class represented by the "representative" document.
             db_dict = {

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1047,35 +1047,64 @@ def materialize_alldocs(context) -> int:
     # Build alldocs
     context.log.info("constructing `alldocs` collection")
 
+    # For each collection, group its documents by their `type` value, transform them, and load them into `alldocs`.
     for collection_name in collection_names:
         context.log.info(
             f"Found {mdb[collection_name].estimated_document_count()} estimated documents for {collection_name=}."
         )
 
-        # For each document in this collection, replace the value of the `type` field with
-        # a _list_ of the document's own class and ancestor classes, remove the `_id` field,
-        # and insert the resulting document into the `alldocs` collection.
-        num_docs_inserted_for_collection = 0
-        for doc in mdb[collection_name].find():
-            # Instantiate the Python class represented by the document.
-            db_dict = {collection_name: [dissoc(doc, "_id")]}
+        # Process all the distinct `type` values (i.e. value in the `type` field) of the documents in this collection.
+        #
+        # References:
+        # - https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html#pymongo.collection.Collection.distinct
+        #
+        distinct_type_values = mdb[collection_name].distinct(key="type")
+        for type_value in distinct_type_values:
+
+            # Process all the documents in this collection that have this value in their `type` field.
+            #
+            # References:
+            # - https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html#pymongo.collection.Collection.count_documents
+            # - https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html#pymongo.collection.Collection.find
+            #
+            filter_ = {"type": type_value}
+            num_docs_having_type = mdb[collection_name].count_documents(filter=filter_)
+            docs_having_type = mdb[collection_name].find(filter=filter_)
+            context.log.info(
+                f"Found {len(num_docs_having_type)} documents having {type_value=} in {collection_name=}."
+            )
+
+            # Get a "representative" document from the result.
+            #
+            # Note: Since all of the documents in this batch have the same class ancestry, we will save time by
+            #       determining the class ancestry of only _one_ of them (we call this the "representative") and then
+            #       (later) attributing that class ancestry to all of them.
+            #
+            representative_doc = next(docs_having_type, default=None)
+
+            # Instantiate the Python class represented by the "representative" document.
+            db_dict = {collection_name: [dissoc(representative_doc, "_id")]}  # omits key incompatible with constructor
             nmdc_db = NMDCDatabase(**db_dict)
-            doc_as_instance = getattr(nmdc_db, collection_name)[0]
+            representative_instance = getattr(nmdc_db, collection_name)[0]
 
-            # Get the ancestry of that instance, as a list of class names (including its own class name).
-            ancestor_class_names = class_hierarchy_as_list(doc_as_instance)
+            # Get the class ancestry of that instance, as a list of class names (including its own class name).
+            ancestor_class_names = class_hierarchy_as_list(representative_instance)
 
-            # Make a document whose `type` field contains that list of class names.
-            # Note: This clobbers the existing contents of the `type` field.
-            new_doc = assoc(dissoc(doc, "type", "_id"), "type", ancestor_class_names)
-
-            # Store this document in the `alldocs` collection.
-            mdb.alldocs.insert_one(new_doc)
-            num_docs_inserted_for_collection += 1
-
-        context.log.info(
-            f"Inserted {num_docs_inserted_for_collection} documents for {collection_name=}."
-        )
+            # Store the documents belonging to this group, in the `alldocs` collection, setting their `type` field
+            # to the list of class names obtained from the "representative" document above.
+            #
+            # TODO: Document why clobbering the existing contents of the `type` field is OK.
+            #
+            inserted_many_result = mdb.alldocs.insert_many(
+                [
+                    assoc(dissoc(doc, "type", "_id"), "type", ancestor_class_names)
+                    for doc in docs_having_type
+                ]
+            )
+            context.log.info(
+                f"Inserted {len(inserted_many_result.inserted_ids)} documents originally having {type_value=} "
+                f"in {collection_name=}."
+            )
 
     # Re-idx for `alldocs` collection
     mdb.alldocs.create_index("id", unique=True)

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1071,7 +1071,7 @@ def materialize_alldocs(context) -> int:
             num_docs_having_type = mdb[collection_name].count_documents(filter=filter_)
             docs_having_type = mdb[collection_name].find(filter=filter_)
             context.log.info(
-                f"Found {len(num_docs_having_type)} documents having {type_value=} in {collection_name=}."
+                f"Found {num_docs_having_type} documents having {type_value=} in {collection_name=}."
             )
 
             # Get a "representative" document from the result.
@@ -1084,8 +1084,9 @@ def materialize_alldocs(context) -> int:
 
             # Instantiate the Python class represented by the "representative" document.
             db_dict = {
+                # Shed the `_id` attribute, since the constructor doesn't allow it.
                 collection_name: [dissoc(representative_doc, "_id")]
-            }  # omits key incompatible with constructor
+            }
             nmdc_db = NMDCDatabase(**db_dict)
             representative_instance = getattr(nmdc_db, collection_name)[0]
 

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1059,6 +1059,9 @@ def materialize_alldocs(context) -> int:
         # - https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html#pymongo.collection.Collection.distinct
         #
         distinct_type_values = mdb[collection_name].distinct(key="type")
+        context.log.info(
+            f"Found {len(distinct_type_values)} distinct `type` values in {collection_name=}: {distinct_type_values=}"
+        )
         for type_value in distinct_type_values:
 
             # Process all the documents in this collection that have this value in their `type` field.

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1083,7 +1083,9 @@ def materialize_alldocs(context) -> int:
             representative_doc = next(docs_having_type, default=None)
 
             # Instantiate the Python class represented by the "representative" document.
-            db_dict = {collection_name: [dissoc(representative_doc, "_id")]}  # omits key incompatible with constructor
+            db_dict = {
+                collection_name: [dissoc(representative_doc, "_id")]
+            }  # omits key incompatible with constructor
             nmdc_db = NMDCDatabase(**db_dict)
             representative_instance = getattr(nmdc_db, collection_name)[0]
 

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1022,46 +1022,59 @@ def materialize_alldocs(context) -> int:
     mdb = context.resources.mongo.db
     collection_names = populated_schema_collection_names_with_id_field(mdb)
 
-    for name in collection_names:
-        assert (
-            len(collection_name_to_class_names[name]) == 1
-        ), f"{name} collection has class name of {collection_name_to_class_names[name]} and len {len(collection_name_to_class_names[name])}"
+    # Insert a no-op as an anchor point for this comment.
+    #
+    # Note: There used to be code here that `assert`-ed that each collection could only contain documents of a single
+    #       type. With the legacy schema, that assertion was true. With the Berkeley schema, it is false. That code was
+    #       in place because subsequent code (further below) used a single document in a collection as the source of the
+    #       class ancestry information of _all_ documents in that collection; an optimization that spared us from
+    #       having to do the same for every single document in that collection. With the Berkeley schema, we have
+    #       eliminated that optimization (since it is inadequate; it would produce some incorrect class ancestries
+    #       for descendants of `PlannedProcess`, for example).
+    #
+    pass
 
     context.log.info(f"{collection_names=}")
 
     # Drop any existing `alldocs` collection (e.g. from previous use of this op).
+    #
+    # FIXME: This "nuke and pave" approach introduces a race condition.
+    #        For example, if someone were to visit an API endpoint that uses the "alldocs" collection,
+    #        the endpoint would fail to perform its job since the "alldocs" collection is temporarily missing.
+    #
     mdb.alldocs.drop()
 
     # Build alldocs
     context.log.info("constructing `alldocs` collection")
 
-    for collection in collection_names:
-        # Calculate class_hierarchy_as_list once per collection, using the first document in list
-        try:
-            nmdcdb = NMDCDatabase(
-                **{collection: [dissoc(mdb[collection].find_one(), "_id")]}
-            )
-            exemplar = getattr(nmdcdb, collection)[0]
-            newdoc_type: list[str] = class_hierarchy_as_list(exemplar)
-        except ValueError as e:
-            context.log.info(f"Collection {collection} does not exist.")
-            raise e
-
+    for collection_name in collection_names:
         context.log.info(
-            f"Found {mdb[collection].estimated_document_count()} estimated documents for {collection=}."
+            f"Found {mdb[collection_name].estimated_document_count()} estimated documents for {collection_name=}."
         )
+
         # For each document in this collection, replace the value of the `type` field with
         # a _list_ of the document's own class and ancestor classes, remove the `_id` field,
         # and insert the resulting document into the `alldocs` collection.
+        num_docs_inserted_for_collection = 0
+        for doc in mdb[collection_name].find():
+            # Instantiate the Python class represented by the document.
+            db_dict = {collection_name: [dissoc(doc, "_id")]}
+            nmdc_db = NMDCDatabase(**db_dict)
+            doc_as_instance = getattr(nmdc_db, collection_name)[0]
 
-        inserted_many_result = mdb.alldocs.insert_many(
-            [
-                assoc(dissoc(doc, "type", "_id"), "type", newdoc_type)
-                for doc in mdb[collection].find()
-            ]
-        )
+            # Get the ancestry of that instance, as a list of class names (including its own class name).
+            ancestor_class_names = class_hierarchy_as_list(doc_as_instance)
+
+            # Make a document whose `type` field contains that list of class names.
+            # Note: This clobbers the existing contents of the `type` field.
+            new_doc = assoc(dissoc(doc, "type", "_id"), "type", ancestor_class_names)
+
+            # Store this document in the `alldocs` collection.
+            mdb.alldocs.insert_one(new_doc)
+            num_docs_inserted_for_collection += 1
+
         context.log.info(
-            f"Inserted {len(inserted_many_result.inserted_ids)} documents for {collection=}."
+            f"Inserted {num_docs_inserted_for_collection} documents for {collection_name=}."
         )
 
     # Re-idx for `alldocs` collection

--- a/nmdc_runtime/site/ops.py
+++ b/nmdc_runtime/site/ops.py
@@ -1105,8 +1105,8 @@ def materialize_alldocs(context) -> int:
                 ]
             )
             context.log.info(
-                f"Inserted {len(inserted_many_result.inserted_ids)} documents originally having {type_value=} "
-                f"in {collection_name=}."
+                f"Inserted {len(inserted_many_result.inserted_ids)} documents from {collection_name=} "
+                f"originally having {type_value=}."
             )
 
     # Re-idx for `alldocs` collection

